### PR TITLE
Warn user when comparing graphs of the same name

### PR DESCRIPTION
**Pull request by aldebaran30701**
### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first. Failed Unit tests can be 
    debugged by adding the label "verbose logging" to the GitHub PR.

### Description of the Change
Compare Graph allows the user to pick from a list of graphs. It is possible for the user to either pick both graphs as the same graph, or pick two graphs wit the same name. Both cases will be confusing and the best course of action seems to be to warn the user of this.

I suggest the following, throw an exception when the compared graphs have the same name. This will force the user to change the name of one of the graphs, which will reduce confusion of the results, but also hint to them that they may have accidentally selected the same graph to compare with itself.

I decided against a "WARNING" level of notification as it did not hint to me that the comparison was not done. An "ERROR" level message seemed to be the best fit.

Below is an image of the error which appears when the names for compared graphs are the same.

image

### Alternate Designs
N/A

### Why Should This Be In Core?
Usability

### Benefits
Usability

### Possible Drawbacks
N/A

### Verification Process
Comparing graphs with the same name, check for error message. Compare graph with different names, check for no error message.

### Applicable Issues
#1350

### CODE REVIEW

**Comment by antares1470** [comment]
@aldebaran30701 While I agree that I don't think you should be able to compare a graph with itself (it doesn't make sense), I'm not sure whether I'm a huge fan of this solution. A couple of reasons why:
1) Given we don't have any issues with the same file name everywhere else in the application (provided the file path is in different for each one), I'm not sure I'm keen on restricting it just here. While it may make the output easier to read, I wonder whether some users may not like the idea of having to change the name of a graph just to make one function work.
2) The solution is reactive when I feel a more proactive solution may be better suited. Rather than flashing a popup at someone telling them they're trying to compare with themselves, it may be better to simply not allow them to pick the same graph twice in the first place. e.g. if one graph is selected in the first drop-down, the 2nd drop-down could be updated to remove that graph from the options. One step further, you could potentially not allow users to run Compare graph at all if there isn't at least a couple of graphs open

**Comment by aldebaran30701** [comment]
My most recent commit adds a controller to the plugin parameter. The issue with this is I've used it directly how I feel Select Top N Plugin is used. I step through with the debugger and the controller seems to set the compare parameter to the same list minus the original graph. Which is what we want. But what actually happens on the app is the dialog still has that original graph. 
Selecting something from either menu throws a null pointer which I cannot resolve where it is actually stemming from. 

```
java.lang.NullPointerException
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.removeListener(ExpressionHelper.java:73)
	at javafx.base/javafx.beans.property.ReadOnlyBooleanPropertyBase.removeListener(ReadOnlyBooleanPropertyBase.java:61)
	at javafx.graphics/javafx.stage.PopupWindow.unbindOwnerFocusedProperty(PopupWindow.java:895)
	at javafx.graphics/javafx.stage.PopupWindow.doVisibleChanged(PopupWindow.java:542)
	at javafx.graphics/javafx.stage.PopupWindow.access$100(PopupWindow.java:102)
	at javafx.graphics/javafx.stage.PopupWindow$1.doVisibleChanged(PopupWindow.java:111)
	at javafx.graphics/com.sun.javafx.stage.PopupWindowHelper.visibleChangedImpl(PopupWindowHelper.java:63)
	at javafx.graphics/com.sun.javafx.stage.WindowHelper.visibleChanged(WindowHelper.java:77)
	at javafx.graphics/javafx.stage.Window$12.invalidated(Window.java:1147)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:110)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:145)
	at javafx.graphics/javafx.stage.Window.setShowing(Window.java:1174)
	at javafx.graphics/javafx.stage.Window.hide(Window.java:1199)
	at javafx.graphics/javafx.stage.PopupWindow.hide(PopupWindow.java:486)
	at javafx.controls/javafx.scene.control.skin.ComboBoxPopupControl.hide(ComboBoxPopupControl.java:258)
	at javafx.controls/javafx.scene.control.skin.ComboBoxBaseSkin.lambda$new$6(ComboBoxBaseSkin.java:130)
	at javafx.controls/com.sun.javafx.scene.control.LambdaMultiplePropertyChangeListenerHandler.lambda$new$1(LambdaMultiplePropertyChangeListenerHandler.java:49)
	at javafx.base/javafx.beans.value.WeakChangeListener.changed(WeakChangeListener.java:86)
[catch] at javafx.base/com.sun.javafx.binding.ExpressionHelper$Generic.fireValueChangedEvent(ExpressionHelper.java:360)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ReadOnlyBooleanPropertyBase.fireValueChangedEvent(ReadOnlyBooleanPropertyBase.java:72)
	at javafx.base/javafx.beans.property.ReadOnlyBooleanWrapper.fireValueChangedEvent(ReadOnlyBooleanWrapper.java:103)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:111)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:145)
	at javafx.controls/javafx.scene.control.ComboBoxBase.setShowing(ComboBoxBase.java:187)
	at javafx.controls/javafx.scene.control.ComboBoxBase.hide(ComboBoxBase.java:404)
	at impl.org.controlsfx.skin.SearchableComboBoxSkin.lambda$createFilteredComboBox$0(SearchableComboBoxSkin.java:188)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper$SingleChange.fireValueChangedEvent(ExpressionHelper.java:181)
	at javafx.base/com.sun.javafx.binding.ExpressionHelper.fireValueChangedEvent(ExpressionHelper.java:80)
	at javafx.base/javafx.beans.property.ReadOnlyBooleanPropertyBase.fireValueChangedEvent(ReadOnlyBooleanPropertyBase.java:72)
	at javafx.base/javafx.beans.property.ReadOnlyBooleanWrapper.fireValueChangedEvent(ReadOnlyBooleanWrapper.java:103)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.markInvalid(BooleanPropertyBase.java:111)
	at javafx.base/javafx.beans.property.BooleanPropertyBase.set(BooleanPropertyBase.java:145)
	at javafx.controls/javafx.scene.control.ComboBoxBase.setShowing(ComboBoxBase.java:187)
	at javafx.controls/javafx.scene.control.ComboBoxBase.hide(ComboBoxBase.java:404)
	at au.gov.asd.tac.constellation.plugins.gui.SingleChoiceInputPane.lambda$new$1(SingleChoiceInputPane.java:107)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$10(PlatformImpl.java:428)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at javafx.graphics/com.sun.javafx.application.PlatformImpl.lambda$runLater$11(PlatformImpl.java:427)
	at javafx.graphics/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:96)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
	at javafx.graphics/com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:174)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by antares1470** [comment]
@aldebaran30701 I did a quick check and issue doesn't appear to be the code you've added but rather the code you haven't removed. The compare graph parameter is still also having its options updated in the `updateParameters` function so removing that resolves your issue (including null pointer)

**Comment by aldebaran30701** [comment]
This should be ready for review - Thanks for that pointer @antares1470 it was right on the money.

File: CoreFunctionality/src/au/gov/asd/tac/constellation/functionality/compare/CompareGraphPlugin.java [hunk 1]
```diff
@@ -227,9 +227,9 @@ public void updateParameters(final Graph graph, final PluginParameters parameter
         SingleChoiceParameterType.setOptions(originalGraph, graphNames);
         SingleChoiceParameterType.setChoice(originalGraph, GraphNode.getGraphNode(graph.getId()).getDisplayName());
 
-        @SuppressWarnings("unchecked") //COMPARE_GRAPH_PARAMETER will always be of type SingleChoiceParameter
-        final PluginParameter<SingleChoiceParameterValue> compareGraph = (PluginParameter<SingleChoiceParameterValue>) parameters.getParameters().get(COMPARE_GRAPH_PARAMETER_ID);
-        SingleChoiceParameterType.setOptions(compareGraph, graphNames);
+//        @SuppressWarnings("unchecked") //COMPARE_GRAPH_PARAMETER will always be of type SingleChoiceParameter
+//        final PluginParameter<SingleChoiceParameterValue> compareGraph = (PluginParameter<SingleChoiceParameterValue>) parameters.getParameters().get(COMPARE_GRAPH_PARAMETER_ID);
+//        SingleChoiceParameterType.setOptions(compareGraph, graphNames);
```
Comment (antares1470) [inline CoreFunctionality/src/au/gov/asd/tac/constellation/functionality/compare/CompareGraphPlugin.java:232]: Are you able to just remove the code (assuming it's not needed)

**Review by antares1470 — CHANGES_REQUESTED** [review]
A suggestion from testing: If possible I think it would be good to not allow people to run the Compare Graph plugin if there isn't at least 2 graphs open. If only one graph is open, there won't be an option for a compare graph to compare against (which is expected) but maybe it shouldn't ever get to that point in the first place (i.e. if only one graph is open and you select the Compare Graph action from the menu, could be better to just present a popup telling the user that at least 2 graphs need to be open to run the plugin)

**Review by antares1470 — APPROVED** [review]

**Review by arcturus2 — APPROVED** [review]

**Pull request merged**